keps/sig-node/kubelet-node-endpoint.md

```diff
@@ -0,0 +1,89 @@
+---
+title: Kubelet Node Endpoint
+authors:
+  - "clamoriniere"
+owning-sig: sig-node
+participating-sigs: []
+reviewers: []
+approvers: []
+creation-date: 2019-05-02
+status: implementable
+---
+
+# Kubelet Node Endpoint
+
+## Motivation
+
+The Kubelet already proposes a set of endpoints (`/pods`, and soon `/metrics/resource/v1*`) that offer useful information for "local" node monitoring. In addition to those endpoints, we would like to introduce a new `/node` endpoint for displaying the node resource information. It will allow daemonset-based monitoring solutions to retrieve node information locally (on the same host) like "labels, annotations" instead of using a call to the API server, which will solve some scalability issues in big clusters.
+
+![kubelet-node-endpoint](https://i.ibb.co/khVg8rd/Untitled-Diagram.png)
+
+## Proposal
+
+Kubelet already caches and sync its corresponding `Node` information from the API-server. The proposed solution opens a new endpoint "/node" that returns the `k8s.io/api/pkg/core/v1.Node` JSON representation of the Kubelet's Node. Only the Kubelet's Node resource is return.
+
+For consistency the endpoint path "/node" looks similar to the "/pods" endpoint (no API versioning).
+
+## Use Cases
+
+* **Pod metrics and logs tagging with Node labels or annotation**: Monitoring workload solutions based on the kubelet API ("/pods" and "/metrics") used to decorate Pod metrics with additional context information; Information like the "Availability zone", "Provider", "Server type" etc are very useful for issue investigation.
+The current workaround is to retrieve that information from the API-Server, that generates a consequent number of requests to the API-Server on a large kubernetes cluster.
```

Comment:
Many monitoring pipelines perform joins to accomplish this kind of correlation.  For example, if a pod metric has the node name attached as a label, a node_info metric that has "Availability zone", "Provider", etc. can be joined with the pod metric as long as both have the node name as a label.  In my experience, a single watch on all nodes shouldn't be prohibitively expensive, especially after the node lease feature is enabled (alpha in 1.13 IIRC).  Granted, joining can also be an expensive operation for query latency or storage if the join is pre-computed.  Does that model solve this use-case for your monitoring setup?

Reply:
We initially designed a solution like you describes.

But we faced some scaling issues; what is costly is not watching ones all nodes, but watching one node in "N" times: one listener on each daemonset agent pod. It generates a large number of connection to the API-server (for example on k8s cluster with 2000 nodes)

Also Node labels like "Availability zone", "Provider" are stable, it didn't increase the cardinality of Pod metrics and so did increase the storage capacity needs. In addition, querying the metrics will be cheaper since no join at query time will be needed.

Reply:
I think we are agreeing about what the trade-offs are here.  I am just pointing out that a **single, cluster-level watcher of nodes** can produce all the info metrics that can be joined with pod metrics without causing scalability problems.  Doing a join at query time makes large queries slower.  Doing a join after both metrics have been collected, but before query time ("precomputed" join) means storing the pod, node, and precomputed pod+node metrics, and thus requires extra storage.  Thus far, existing monitoring solutions, such as prometheus, have taken this approach (e.g. join with metrics from kube-state-metrics).  This is nice because it separates the collection of pod and node metrics, and is generally more flexible, as I can determine which metrics to join further up the stack.

What i'm trying to figure out is why you prefer joining node and pod information in the per-node agent vs doing a join after metric collection as other pipelines have chosen to do.  I'm not suggesting you must do it one way or another, just trying to understand your constraints.

Reply:
having one single cluster-level watcher of nodes is what we are currently using (cf: https://www.datadoghq.com/blog/datadog-cluster-agent/) it solves part of the problem, by limiting connection to the API-server.

But in a monitoring solution base on agent deploy on each node, metrics don't transit by a "proxy" which can attach to the metrics the node informations. each agent send complete metrics (with already most of the labels) directly to the centralise metrics store pipeline.

I know how prometheus is joining labels like the nodes/namespace labels on metrics after collecting them before storing them in the TSDB; and in a sens it is what we want to achieve, but in our case in each node agent instance, before sending metrics to our metrics backend services.